File: suffixarray/suffixarray.js
var SuffixArray = function(str) {
  var sa = [];
  var i;
  var l = str.length;
  for (i = 0; i < l; ++i) {
    sa.push(i);
  }
  sa.sort(function(a, b) {
    return str.substr(a) < str.substr(b) ? -1 : 1;
  });
  return sa;
};

exports.SuffixArray = SuffixArray;

function binarySearch(str, sa, key){
  var low = 0;
  var high = sa.length - 1;
  var mid;
  while(low <= high) {
    mid = Math.floor((low + high) / 2);
    var val = str.substr(sa[mid], key.length);
    //console.log("(mid, key, val): " + mid + ", " + key + ", " + val);
    if(val === key) {
      return mid;
    } else if(val < key) {
      low = mid + 1;
    } else {
      high = mid - 1;
    }
  }
  return null;
}

exports.binarySearch = binarySearch;

File: suffixarray/suffixarray.py
#-*- coding: utf-8 -*-
import math

def make_suffix_array(s):
    l = len(s)
    sa = list(range(l))
    sa.sort(cmp=lambda a,b: -1 if s[a] < s[b] else 1)
    return sa

def binarySearch(s, sa, key):
    low  = 0
    high = len(sa) - 1;
    while low <= high:
        mid = int(math.floor((low + high) / 2))
        val = s[sa[mid]:sa[mid] + len(key)]
        #print("(mid, key, val): %s, %s, %s" % (mid, key, val))
        if val == key:
            return mid
        elif val < key:
            low = mid + 1
        else:
            high = mid - 1
    return None

if __name__ == '__main__':
    s    = 'abracadabra'
    word = 'racadabra'
    s    = u'ぁっぉトイプー'
    word = u'ぉ'
    sa = make_suffix_array(s)
    r = binarySearch(s, sa, word)
    if r: print(s[sa[r]:sa[r]+len(word)])
